File: web/src/utils/syncEntityUtils.ts
import type { Category, Expense } from '@/models/types';

export function newCloudId(): string {
  return crypto.randomUUID();
}

export function ensureCloudId<T extends { cloudId?: string }>(item: T): T & { cloudId: string } {
  return { ...item, cloudId: item.cloudId?.trim() || newCloudId() };
}

export function stampedForSync<T extends { cloudId?: string; updatedAt?: number; pendingSync?: boolean }>(
  item: T,
): T & { cloudId: string; updatedAt: number; pendingSync: true } {
  const withId = ensureCloudId(item);
  return {
    ...withId,
    updatedAt: Date.now(),
    pendingSync: true,
  };
}

export function stampedCategory(category: Omit<Category, 'id' | 'cloudId' | 'updatedAt' | 'pendingSync'> & Partial<Pick<Category, 'cloudId'>>): Category {
  return stampedForSync(category as Category);
}

export function stampedExpense(expense: Omit<Expense, 'id' | 'cloudId' | 'updatedAt' | 'pendingSync'> & Partial<Pick<Expense, 'cloudId'>>): Expense {
  return stampedForSync(expense as Expense);
}

File: app/src/main/java/com/aus/ausgegeben/data/SyncEntityUtils.kt
package com.aus.ausgegeben.data

import com.aus.ausgegeben.data.entity.Category
import com.aus.ausgegeben.data.entity.Expense
import java.util.UUID

fun Category.ensureCloudId(): Category =
    if (cloudId.isBlank()) copy(cloudId = UUID.randomUUID().toString()) else this

fun Category.stampedForSync(): Category =
    ensureCloudId().copy(updatedAt = System.currentTimeMillis(), pendingSync = true)

fun Expense.ensureCloudId(): Expense =
    if (cloudId.isBlank()) copy(cloudId = UUID.randomUUID().toString()) else this

fun Expense.stampedForSync(): Expense =
    ensureCloudId().copy(updatedAt = System.currentTimeMillis(), pendingSync = true)
